tfx/experimental/pipeline_testing/pipeline_recorder_utils.py

```diff
@@ -0,0 +1,150 @@
+  """Fetches executions associated with the latest context.
+
+  Args:
+    metadata_connection: A class for metadata I/O to metadata db.
+
+  Returns:
+    A dictionary that holds executions for pipeline run_id
+  """
+  pipeline_run_contexts = [
+      c for c in metadata_connection.store.get_contexts_by_type(
+          metadata._CONTEXT_TYPE_PIPELINE_RUN)  # pylint: disable=protected-access
+      if c.properties['pipeline_name'].string_value == pipeline_name
+  ]
+  latest_context = max(pipeline_run_contexts, key=lambda c: c.last_update_time_since_epoch)
+  return metadata_connection.store.get_executions_by_context(latest_context.id)
+
+def record_pipeline(output_dir: Text,
+                    metadata_db_uri: Optional[Text],
+                    host: Optional[Text],
+                    port: Optional[int],
+                    pipeline_name: Optional[Text],
+                    run_id: Optional[Text]) -> None:
+  """Record pipeline run with run_id to output_dir. For the beam pipeline,
+  metadata_db_uri is required. For KFP, host and port should be specified.
+
+  Args:
+    output_dir: Directory path to which pipeline outputs are recorded.
+    metadata_db_uri: Uri to metadata db.
+    host: The host to connect to gRPC server.
+    port: The port to connect to gRPC server.
+    run_id: Pipeline execution run_id.
+
+  Raises:
+    ValueError: In cases of invalid arguments:
+      - metadata_db_uri is None or host and/or port is None.
+      - run_id is None and pipeline_name is None.
+  """
+  if host is not None and port is not None:
+    metadata_config = metadata_store_pb2.MetadataStoreClientConfig()
+    metadata_config.host = host
+    metadata_config.port = port
+  elif metadata_db_uri is not None:
+    metadata_config = metadata.sqlite_metadata_connection_config(
+        metadata_db_uri)
+  else:
+    raise ValueError("For KFP, host and port are required. "\
+                     "For beam pipeline, metadata_db_uri is required.")
+
+  with metadata.Metadata(metadata_config) as metadata_connection:
+    if run_id is None:
+      if pipeline_name is None:
+        raise ValueError("If the run_id is not specified,"\
+                         " pipeline_name should be specified")
+      # fetch executions of the most recently updated execution context.
+      executions = get_latest_executions(metadata_connection,
+                                         pipeline_name)
+    else:
+      execution_dict = get_execution_dict(metadata_connection)
+      if run_id in execution_dict:
+        executions = execution_dict[run_id]
+      else:
+        raise ValueError(
+            "run_id {} is not recorded in the MLMD metadata".format(run_id))
+    for src_uri, dest_uri in \
+          get_paths(metadata_connection, executions, output_dir):
+      if not tf.io.gfile.exists(src_uri):
+        raise FileNotFoundError("{} does not exist".format(src_uri))
+      os.makedirs(dest_uri, exist_ok=True)
```

Comment:
We don't need to make directories here. (And os.makedirs won't work against gs:// paths.) io_utils.copy_dir will make directories for the destination.